RFC-0006-nan-definition.md

```diff
@@ -0,0 +1,179 @@
+>….							
+>Operations on numbers are performed according to the normal rules of arithmetic, within implementation- defined limits, except as provided for in Subclause 6.29, “<numeric value expression>”. 
+
+Some read the above as disallowing NaNs because floating point types are defined only as mantissa and exponent and there are no special values defined like NaN, Infinity, and -Infinity.  I don't think it's clear one way or the other, and it looks like most dbs do support NaN values.
+
+With regard to sorting, the spec says the following:							>											
+> f)  PVi is said to precede QVi if the value of the <comparison predicate> “PVi <comp op> QVi” is True for the applicable <comp op>.			
+>			 							
+> g)  If PVi and QVi are not the null value and the result of “PVi <comp op> QVi” is Unknown, then the relative ordering of PVi and QVi is implementation-dependent.
+
+### Current Presto behavior
+
+Here we give a short summary of Presto's current behavior.  There is a complete review here: https://docs.google.com/spreadsheets/d/1KclsskH88CLHMiu_Q2P25S3QfPc9H3Tu7h_9MXP10Dk/edit#gid=128250498  
+
+Currently basic comparison operators follow IEEE-754. =  returns false. <> returns true.  Joins on NaN keys do not match.  For maps and arrays, NaNs are considered distinct from each other, and will appear multiple times in a set_agg, union, or even as map_keys.  However, for group bys, distinct, and distinct aggregations, NaNs are deduplicated (there will only be one group for all NaNs, they are not considered distinct from each other). 
+
+For sorting, again the basic comparisons follow IEEE-754. where <,>, <=, >= return false.  Beyond that there is a considerable degree of variation across functions and operators. min, max, min_by, max_by, min(x, n) and max(x,n) have inconsistent behavior depending on when NaN is encountered.  Greatest and Least throw an error on NaN input. Map_top_n returns wrong results if NaN shows up in the input. Order by sorts NaN as larger than infinity, and most array and map sorting and topn functions consider NaN as largest as well. However, array_min and array_max return NaN if any of the input is NaN.
+
+When it comes to pushing filters into the scan. Tuple domains, I believe would consider NaN largest (though needs a bit more review to be certain). tupleDomainFilters testDouble() will return false if the value is NaN unless the filter is a NOT IN (following the IEEE definition and Presto's behavior for basic operators).  Orc files, at least as written by Presto and Velox won't include min/max stats if any of the values are NaN.  Parquet files written by Presto, I believe will consider NaN largest in their min/max stats, but other parquet writers behave differently, and Presto may read files written by other engines.  
```

Comment:
I wonder if it's more desirable for NaN to be excluded from min/max stats for both ORC and Parquet?  One NaN seems to render the stat useless, at least in Parquet: https://github.com/apache/parquet-format/pull/88

Reply:
to exclude NaN or to exclude min/max stats if there are NaNs? i think for the file format we should follow whatever the file format expects if its defined and then make sure to handle it appropriately in Presto.